### PR TITLE
[AutoDiff upstream] add more validation tests

### PR DESCRIPTION
**Pull request by marcrasi**
These are copied straight from the `tensorflow` branch without any modifications necessary.

### CODE REVIEW

**Review by rxwei — APPROVED** [review]

**Comment by marcrasi** [comment]
@swift-ci please test

**Review by dan-zheng — APPROVED** [review]

**Comment by marcrasi** [comment]
Oh no, there is an `import Glibc` causing a failure on `iphonesimulator-i386`. I shall try to fix it using [this tip](https://github.com/apple/swift/pull/30915#discussion_r405983213).

**Comment by marcrasi** [comment]
@swift-ci please test

**Comment by marcrasi** [comment]
@swift-ci please test

**Comment by marcrasi** [comment]
@swift-ci please test windows

File: test/AutoDiff/validation-test/method.swift [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+) {
+  func f(_ p: Parameter) -> Tracked<Float> {
+    return 100 * Parameter.squared(p: p)
+  }
+  expectEqual(Parameter(x: 4 * 100), gradient(at: Parameter(x: 2), in: f))
+  expectEqual(Parameter(x: 40 * 100), gradient(at: Parameter(x: 20), in: f))
+}
+
+// TODO(SR-8699): Fix this test.
+// MethodTests.testWithLeakChecking(
```
Comment (CodaFi) [inline test/AutoDiff/validation-test/method.swift:127]: I believe `StdlibUnitTest` has a nice built-in way of XFAILing tests. Does this not consistently fail/pass/crash?

File: test/AutoDiff/validation-test/method.swift [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+MethodTests.testWithLeakChecking("static method with custom adjoint, called from differentated func") {
+  func f(_ p: CustomParameter) -> Tracked<Float> {
+    return 100 * CustomParameter.squared(p: p)
+  }
+  expectEqual(CustomParameter(x: 4 * 100), gradient(at: CustomParameter(x: 2), in: f))
+  expectEqual(CustomParameter(x: 10 * 100), gradient(at: CustomParameter(x: 20), in: f))
+}
+
+// TODO(SR-8699): Fix this test.
+// MethodTests.testWithLeakChecking("static method with custom adjoint, differentiated directly") {
```
Comment (CodaFi) [inline test/AutoDiff/validation-test/method.swift:341]: Same here.

File: test/AutoDiff/validation-test/superset_adjoint.swift [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    return apply(x, y)
+  }
+  expectEqual(1, gradient(at: Tracked<Float>(0)) { x in foo(x, 0) { $0 + $1 } })
+}
+
+// FIXME: The expression `(+) as @differentiable (Float, @noDerivative Float) -> Float)`
+// forms a curry thunk of `Float.+` before conversion to @differentiable, and AD
+// doesn't know how to differentiate the curry thunk, so it produces a
+// "function is not differentiable" error.
+// SupersetVJPTests.test("CrossModule") {
```
Comment (CodaFi) [inline test/AutoDiff/validation-test/superset_adjoint.swift:62]: This one can be XFailed explicitly.

**Comment by marcrasi** [comment]
Thanks for catching those commented tests. They actually all pass now, so I have uncommented them.

**Comment by marcrasi** [comment]
@swift-ci please test

**Comment by marcrasi** [comment]
@swift-ci please test windows

**Pull request merged**